### PR TITLE
[Feat] Add trip layer - 1 (#632)

### PR DESCRIPTION
**Pull request by yanhann10**
- Add trip layer
- Time slider playback
- Trip layer icon

### CODE REVIEW

File: src/utils/layer-utils/layer-utils.js [hunk 1]
```diff
@@ -35,17 +36,26 @@ export function findDefaultLayer(dataset, layerClasses) {
   let layers = [];
   Object.keys(layerClasses).forEach(lc => {
     const layerProps = layerClasses[lc].findDefaultLayerProps(dataset);
```
Comment (heshan0131) [inline src/utils/layer-utils/layer-utils.js]:
```
if (!layerClasses[lc]. findDefaultLayerProps) {
  return;
}
```

File: src/layers/trip-layer/trip-layer.js [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+        (typeof label === 'string' && label.replace(/\.[^/.]+$/, '')) ||
+        this.type,
+      columns,
+      isVisible: true
+    }));
+  }
+
+  getDefaultLayerConfig(props = {}) {
+    return {
+      ...super.getDefaultLayerConfig(props),
```
Comment (heshan0131) [inline src/layers/trip-layer/trip-layer.js]: set `animation: {enabled: true}`

File: src/utils/layer-utils/layer-utils.js [hunk 3]
```diff
@@ -94,3 +104,38 @@ export function getLightSettingsFromBounds(bounds) {
...
+
+// Detect if the data is trip-layer-animatable
+// TODO: to be refined to accommodate more scenarios of input format
+export function tripLayerAnimatable(dataset) {
+  const dataContent = dataset.allData;
+  return dataset && !dataContent.map(d => d[0].geometry).some(e => !e);
+}
+
+// Get time animation domain for a dataset
+export function getTimeAnimationDomainPerDataset(dataset, timestampIndex) {
```
Comment (heshan0131) [inline src/utils/layer-utils/layer-utils.js]: rename it to 
```
function getTimeAnimationDomainFoTripLayer(layer) {
  const {columns, dataId} = this.config;
  const dataContent = datasets[dataId].allData;
  const geojsonFieldIdx = columns.geojson.fieldIdx;

  const timeField = dataContent
    .map(d => d[geojsonFieldIdx].geometry.coordinates.map(coord => coord[timestampIndex]))
    .flat();
...
}
```

File: src/utils/layer-utils/layer-utils.js [hunk 4]
```diff
@@ -94,3 +104,38 @@ export function getLightSettingsFromBounds(bounds) {
...
+export function getTimeAnimationDomain(datasets, timestampIndex) {
+  const minRange = [];
+  const maxRange = [];
+  Object.values(datasets).map(d => {
+    if (tripLayerAnimatable(d)) {
+      minRange.push(getTimeAnimationDomainPerDataset(d, timestampIndex)[0]);
+      maxRange.push(getTimeAnimationDomainPerDataset(d, timestampIndex)[1]);
+    }
+  });
+  const mmin = min(minRange);
```
Comment (heshan0131) [inline src/utils/layer-utils/layer-utils.js]: use `extent` from `d3-array`

File: src/reducers/vis-state-updaters.js [hunk 5]
```diff
@@ -618,11 +635,82 @@ export const toggleFilterAnimationUpdater = (state, action) => ({
...
+ * Enable animation domain with the min and max of timestamps from geojson
+ * @memberof visStateUpdaters
+ * @param {Object} state `visState`
+ * @param {Object} action action
+ * @returns {Object} nextState
+ * @public
+ *
+ */
+
+export const enableLayerAnimationUpdater = (state, action) => {
```
Comment (heshan0131) [inline src/reducers/vis-state-updaters.js]: pass 1 layer to this function

File: src/reducers/vis-state-updaters.js [hunk 6]
```diff
@@ -618,11 +635,82 @@ export const toggleFilterAnimationUpdater = (state, action) => ({
...
+  const {oldLayer, datasets} = action;
+  // if (checkGeoJsonHasTs(datasets)) {
+  const [minTs, maxTs] = getTimeAnimationDomain(datasets, 3);
```
Comment (heshan0131) [inline src/reducers/vis-state-updaters.js]: pass that layer to `getTimeAnimationDomain`

File: src/utils/layer-utils/layer-utils.js [hunk 7]
```diff
@@ -94,3 +104,38 @@ export function getLightSettingsFromBounds(bounds) {
...
+    .flat();
+
+  const minTime = min(timeField);
+  const maxTime = max(timeField);
+  return [minTime, maxTime];
+}
+
+// Get time animation domain for multiple datasets
+// When multiple trip-layer-animatable data co-exist, take the min of min and max of max in the domain of each
+export function getTimeAnimationDomain(datasets, timestampIndex) {
```
Comment (heshan0131) [inline src/utils/layer-utils/layer-utils.js]: take 1 layer, by default use `getTimeAnimationDomainFoTripLayer`

```
export function getTimeAnimationDomain(layer){
  return getTimeAnimationDomainFoTripLayer(layer)
}
```

File: src/components/bottom-widget.js [hunk 8]
```diff
@@ -21,57 +21,74 @@
...
 import PropTypes from 'prop-types';
 import TimeWidgetFactory from './filters/time-widget';
+import AnimationControlFactory from './common/animation-control/animation-slider';
+import {WidgetContainer} from 'components/common/styled-components';
 
 const propTypes = {
   filters: PropTypes.arrayOf(PropTypes.object),
   datasets: PropTypes.object,
   uiState: PropTypes.object,
+  visState: PropTypes.object,
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/bottom-widget.js]: fixed

File: src/components/common/animation-control/animation-slider.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
+import React, {Component} from 'react';
+import styled from 'styled-components';
+import {createSelector} from 'reselect';
+import moment from 'moment';
+
+import Slider from 'components/common/slider/slider';
+import {WidgetContainer} from 'components/common/styled-components';
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/common/animation-control/animation-slider.js]: fixed

File: src/components/common/animation-control/animation-slider.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+`;
+
+const IconButton = styled(Button)`
+  padding: 6px 4px;
+  svg {
+    margin: 0 6px;
+  }
+`;
+
+const isUnixTs = 1;
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/common/animation-control/animation-slider.js]: fixed

File: src/components/side-panel/layer-manager.js [hunk 11]
```diff
@@ -1,6 +1,6 @@
 // Copyright (c) 2019 Uber Technologies, Inc.
 //
-// Permission is hereby granted, free of charge, to any person obtaining a copy
+// Permission is hereby granted, free of charge, to anthis.y person obtaining a copy
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-manager.js]: fixed

File: src/components/side-panel/layer-panel/layer-configurator.js [hunk 12]
```diff
@@ -678,6 +679,170 @@ export default class LayerConfigurator extends Component {
...
+        >
+          <VisConfigSlider
+            {...LAYER_VIS_CONFIGS.trailLength}
+            {...visConfiguratorProps}
+            label={false}
+          />
+          )
+        </LayerConfigGroup>
+
+        {/* Elevation */}
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-panel/layer-configurator.js]: fixed

File: src/components/side-panel/layer-panel/layer-configurator.js [hunk 13]
```diff
@@ -678,6 +679,170 @@ export default class LayerConfigurator extends Component {
...
+              />
+              <VisConfigSwitch
+                {...visConfiguratorProps}
+                {...LAYER_VIS_CONFIGS.wireframe}
+              />
+            </ConfigGroupCollapsibleContent>
+          </LayerConfigGroup>
+        ) : null}
+
+        {/* Radius */}
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-panel/layer-configurator.js]: fixed

File: src/components/side-panel/layer-panel/layer-panel.js [hunk 14]
```diff
@@ -127,17 +135,22 @@ function LayerPanelFactory() {
...
-              datasets={datasets}
-              layerTypeOptions={layerTypeOptions}
-              openModal={this.props.openModal}
-              updateLayerConfig={this.updateLayerConfig}
-              updateLayerVisualChannelConfig={this.updateLayerVisualChannelConfig}
-              updateLayerType={this.updateLayerType}
-              updateLayerTextLabel={this.updateLayerTextLabel}
-              updateLayerVisConfig={this.updateLayerVisConfig}
-            />
+            <div>
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-panel/layer-panel.js]: fixed

File: src/layers/base-layer.js [hunk 15]
```diff
@@ -720,19 +739,14 @@ export default class Layer {
...
+      ? FIELD_OPTS[this.config[field].type].scale[channelScaleType]
+      : [this.getDefaultLayerConfig()[scale]];
   }
 
   updateLayerVisualChannel(dataset, channel) {
     const visualChannel = this.visualChannels[channel];
-
     this.validateVisualChannel(channel);
-      // calculate layer channel domain
-    const updatedDomain = this.calculateLayerDomain(dataset, visualChannel);
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/base-layer.js]: fixed

File: src/layers/point-layer/point-layer.js [hunk 16]
```diff
@@ -71,7 +71,7 @@ export const pointVisConfigs = {
   }
 };
 
-export default class PointLayer extends Layer {
+export default class getLineColorPointLayer extends Layer {
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/point-layer/point-layer.js]: fixed

File: src/layers/trip-layer/geojson-utils.js [hunk 17]
```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/trip-layer/geojson-utils.js]: fixed

File: src/utils/layer-utils/layer-utils.js (hunk 4 above)
Comment (yanhann10) [reply src/utils/layer-utils/layer-utils.js]: fixed

File: src/processors/data-processor.js [hunk 18]
```diff
@@ -398,7 +398,7 @@ export function processRowObject(rawData) {
  * 			"type" : "Point",
  * 			"coordinates" : [ -71.073283, 42.417500 ]
  * 		}
- * 	}]
+ * 	}]same
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/processors/data-processor.js]: fixed

File: src/actions/vis-state-actions.js [hunk 19]
```diff
@@ -317,6 +317,35 @@ export function updateAnimationSpeed(idx, speed) {
...
+ * Change filter animation speed
+ * @memberof visStateActions
+ * @param {Number} val -  Current value of the slider
+ * @returns {{type: ActionTypes.PLAY_ANIMATION, val: val}}
+ * @public
+ */
+export function playAnimation(val) {
+  return {
+    type: ActionTypes.PLAY_ANIMATION,
+    val
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/actions/vis-state-actions.js]: done

File: src/components/bottom-widget.js [hunk 20]
```diff
@@ -21,57 +21,74 @@
 import React from 'react';
 import PropTypes from 'prop-types';
 import TimeWidgetFactory from './filters/time-widget';
+import AnimationControlFactory from './common/animation-control/animation-slider';
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/bottom-widget.js]: done

File: src/components/filters/time-widget.js [hunk 21]
```diff
@@ -136,45 +117,47 @@ export class TimeWidget extends Component {
   };
 
   _toggleSpeedControl = () => {
-    this.setState({showSpeedControl: !this.state.showSpeedControl})
+    this.setState({showSpeedControl: !this.state.showSpeedControl});
   };
 
-  fieldSelector = props => props.fields;
-  yAxisFieldsSelector = createSelector(this.fieldSelector, fields =>
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/filters/time-widget.js]: done

File: src/components/filters/time-widget.js [hunk 22]
```diff
@@ -136,45 +117,47 @@ export class TimeWidget extends Component {
   };
 
   _toggleSpeedControl = () => {
-    this.setState({showSpeedControl: !this.state.showSpeedControl})
+    this.setState({showSpeedControl: !this.state.showSpeedControl});
   };
 
-  fieldSelector = props => props.fields;
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/filters/time-widget.js]: done

File: src/components/side-panel.js [hunk 23]
```diff
@@ -152,15 +152,17 @@ export default function SidePanelFactory(
...
         layerTypeChange: visStateActions.layerTypeChange,
         layerVisConfigChange: visStateActions.layerVisConfigChange,
         updateLayerBlending: visStateActions.updateLayerBlending,
         updateLayerOrder: visStateActions.reorderLayer,
         showDatasetTable: this._showDatasetTable,
         showAddDataModal: this._showAddDataModal,
         removeLayer: visStateActions.removeLayer,
-        removeDataset: this._removeDataset
+        removeDataset: this._removeDataset,
+        enlargeFilter: visStateActions.enlargeFilter,
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel.js]: done

File: src/components/side-panel/layer-panel/layer-panel.js [hunk 24]
```diff
@@ -100,6 +104,10 @@ function LayerPanelFactory() {
       this.props.removeLayer(this.props.idx);
     };
 
+    _initiatePlayback = () => {
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-panel/layer-panel.js]: done

File: src/layers/trip-layer/trip-layer.js [hunk 25]
```diff
@@ -0,0 +1,448 @@
...
+
+import Layer, {colorMaker} from '../base-layer';
+import {TripsLayer as DeckGLTripsLayer} from 'deck.gl';
+
+import {hexToRgb} from 'utils/color-utils';
+import {
+  getGeojsonDataMaps,
+  getGeojsonBounds,
+  featureToDeckGlGeoType
+} from './geojson-utils';
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/trip-layer/trip-layer.js]: done

File: src/utils/layer-utils/layer-utils.js (hunk 7 above)
Comment (yanhann10) [reply src/utils/layer-utils/layer-utils.js]: done

File: src/components/bottom-widget.js [hunk 26]
```diff
@@ -21,57 +21,74 @@
...
-    const isAnyFilterAnimating = filters.some(f => f.isAnimating);
-    const enlargedFilterWidth = isOpen ? containerW - sidePanelWidth : containerW;
+    const enlargedFilterWidth = isOpen
+      ? containerW - sidePanelWidth
+      : containerW;
 
-    if (enlargedFilterIdx < 0) {
-      return null;
-    }
+    const animatedLayer = layers.find(l => l.type === 'trip');
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/bottom-widget.js]: done

File: src/components/bottom-widget.js [hunk 27]
```diff
@@ -21,57 +21,74 @@
...
+    const tripLayerVisible = layers
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/bottom-widget.js]: done

File: src/components/side-panel/layer-panel/layer-configurator.js [hunk 28]
```diff
@@ -678,6 +679,170 @@ export default class LayerConfigurator extends Component {
...
+    layerChannelConfigProps
+  }) {
+    const {
+      meta: {featureTypes = {}},
+      config: {visConfig}
+    } = layer;
+
+    return (
+      <StyledLayerVisualConfigurator>
+        {/* Fill Color */}
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/components/side-panel/layer-panel/layer-configurator.js]: done

File: src/layers/trip-layer/trip-layer.js [hunk 29]
```diff
@@ -0,0 +1,448 @@
...
+    type: 'number',
+    defaultValue: 0.5,
+    label: 'Stroke Width',
+    isRanged: false,
+    range: [0, 100],
+    step: 0.1,
+    group: 'stroke',
+    property: 'thickness'
+  },
+  strokeColor: 'strokeColor',
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/trip-layer/trip-layer.js]: done

File: src/utils/layer-utils/layer-utils.js (hunk 3 above)
Comment (yanhann10) [reply src/utils/layer-utils/layer-utils.js]: done

File: src/layers/trip-layer/trip-layer.js [hunk 30]
```diff
@@ -0,0 +1,448 @@
...
+      isVisible: true
+    }));
+  }
+
+  getDefaultLayerConfig(props = {}) {
+    return {
+      ...super.getDefaultLayerConfig(props),
+
+      // add height visual channel
+      heightField: null,
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/trip-layer/trip-layer.js]: done

File: src/layers/trip-layer/trip-layer.js [hunk 31]
```diff
@@ -0,0 +1,448 @@
...
+      getLineWidth: d =>
+        sScale
+          ? this.getEncodedChannelValue(
+              sScale,
+              allData[d.properties.index],
+              sizeField,
+              0
+            )
+          : d.properties.lineWidth || 1,
+      getElevation: d =>
```
[the remark this answers is not in the document]
Comment (yanhann10) [reply src/layers/trip-layer/trip-layer.js]: removed

**Pull request merged**